newton.go

```diff
@@ -0,0 +1,145 @@
+// positive definite. Note that the repeated trial factorization of the
+// modified Hessian involved in this process can be computationally expensive.
+//
+// If the Hessian matrix cannot be formed explicitly or if the computational
+// cost of its factorization is prohibitive, BFGS or L-BFGS quasi-Newton method
+// can be used instead.
+type Newton struct {
+	// LinesearchMethod is a method used for selecting suitable steps along the
+	// descent direction d. Steps should satisfy at least one of the Wolfe,
+	// Goldstein or Armijo conditions. If LinesearchMethod == nil, an
+	// appropriate default is chosen.
+	LinesearchMethod LinesearchMethod
+	// Increase is the factor by which a scalar tau is successively increased
+	// so that (H + tau*I) is positive definite. Larger values reduce the
+	// number of trial Hessian factorizations, but also reduce the second-order
+	// information in H.
+	// Increase must be greater than 1. If Increase is 0, it is defaulted to 5.
+	Increase float64
+
+	linesearch *Linesearch
+
+	hess *mat64.SymDense // Storage for a copy of the Hessian matrix.
+	chol *mat64.TriDense // Storage for the Cholesky factorization.
+	tau  float64
+}
+
+func (n *Newton) Init(loc *Location, f *FunctionInfo, xNext []float64) (EvaluationType, IterationType, error) {
+	if n.Increase == 0 {
+		n.Increase = 5
+	}
+	if n.Increase <= 1 {
+		panic("optimize: Newton.Increase must be greater than 1")
+	}
+	if n.LinesearchMethod == nil {
+		n.LinesearchMethod = &Bisection{}
+	}
+	if n.linesearch == nil {
+		n.linesearch = &Linesearch{}
+	}
+	n.linesearch.Method = n.LinesearchMethod
+	n.linesearch.NextDirectioner = n
+
+	return n.linesearch.Init(loc, f, xNext)
+}
+
+func (n *Newton) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
+	return n.linesearch.Iterate(loc, xNext)
+}
+
+func (n *Newton) InitDirection(loc *Location, dir []float64) (stepSize float64) {
+	dim := len(loc.X)
+	n.chol = resizeTriDense(n.chol, dim)
+	n.hess = resizeSymDense(n.hess, dim)
+	n.tau = 0
+	return n.NextDirection(loc, dir)
+}
+
+func (n *Newton) NextDirection(loc *Location, dir []float64) (stepSize float64) {
+	// This method implements Algorithm 3.3 (Cholesky with Added Multiple of
+	// the Identity) from Nocedal, Wright (2006), 2nd edition.
+
+	dim := len(loc.X)
+	n.hess.CopySym(loc.Hessian)
+
+	// Find the smallest diagonal entry of the Hesssian.
+	minA := n.hess.At(0, 0)
+	for i := 1; i < dim; i++ {
+		a := n.hess.At(i, i)
+		if a < minA {
+			minA = a
+		}
+	}
+	// If the smallest diagonal entry is positive, the Hessian may be positive
+	// definite, and so first attempt to apply the Cholesky factorization to
+	// the un-modified Hessian. If the smallest entry is negative, use the
+	// final tau from the last iteration if regularization was needed,
+	// otherwise guess an appropriate value for tau.
+	if minA > 0 {
+		n.tau = 0
+	} else if n.tau == 0 {
+		n.tau = -minA + 0.001
+	}
+
+	for {
```

Comment:
Yea, we need to do something. If the Hessian has an "inf" for example it will never conclude. Trying a few times and then doing the negative gradient seems reasonable. The other option is to return an error, but most users would probably rather find the optimum.